### DISCUSSION TITLE
`defer` in for loop works unexpectedly

### DISCUSSION THREAD

**Discussion by i582** [vlang/v#17313] [General]
### Describe the bug

Code: https://play.vlang.io/p/def2a9db0f

```v
fn main() {
	for i := 0; i < 10; i++ {
		defer {
			println(i)
		}
	}
}

```

### Expected Behavior

Output:

```
9
8
7
6
...
0
```

See also Go implementation and behavior: https://go.dev/play/p/Z6udIdHNKn6

### Current Behavior

Output:

```
10
```

### Reproduction Steps

Run code above

### V version

0.3.3

### Environment details (OS name and version, etc.)

macOS 13

**Comment by JalonSolov** [comment]
`defer` is working as expected (and documented) for V.

First line in the description:  https://github.com/vlang/v/blob/master/doc/docs.md#defer

`A defer statement defers the execution of a block of statements until the surrounding function returns.`

If you think it should work differently, that would be a feature request.

**Comment by Delta456** [comment]
> `defer` is working as expected (and documented) for V.
> 
> First line in the description: https://github.com/vlang/v/blob/master/doc/docs.md#defer
> 
> `A defer statement defers the execution of a block of statements until the surrounding function returns.`
> 
> If you think it should work differently, that would be a feature request.

I believe defer has to work differently for `loops` like it does in Go.

**Comment by MatejMagat305** [comment]
solution would be forbit defer in loop ..., because if I understand defer is translated statically in comp time .... (defer in golang is dynamic and save to stack ....)

**Comment by Wajinn** [comment]
I think the idea that V should work exactly as Go, is a bit dangerous. There are significant differences between the languages, thus a lot of care should probably be taken. After developers discuss it, then whatever team V decides, is the better direction. If that direction is different from Go, then that should be considered OK too, as two different languages.

**Reply by i582** [reply]
> I think the idea that V should work exactly as Go, is bit dangerous. There are significant differences between the languages, thus a lot of care should probably be taken. After developers discuss it, then whatever team V decides, is the better direction. If that direction is different from Go, then that can be OK too, as two different languages.

All languages work in a similar way in one way or another, I don’t see anything dangerous in the fact that V will take some behavior from Go. Moreover, given that potential users will come from Go, having different behavior in such things can only be confusing (how it was for me)

**Reply by Wajinn** [reply]
Sorry about that move, as attempted to move it as a separate response to what matejmagat305 gave. 

I don't think there is anything wrong with following something that Go has done, but do think that it should be considered carefully and that team V should not feel obligated or forced to go down the same direction. It should arguably be what is best for V. Maybe the same as Go, others times not.

> Moreover, given that potential users will come from Go, having different behavior in such things can only be confusing (how it was for me)

But V is already significantly different from Go in various ways.  While the languages are somewhat similar, we can't expect them to be the same, and do exactly as the other does.  I'm just saying that certain things should be considered in a case by case way.
